README.md
gInfinity
=========

gInfinity extends your favorite search engine Google with infinite scroll. Scroll down to the bottom of the search result and the following 20 search results are automatically added to the search result.

changelog.txt
2.4.1
- Fixed bug in auto tab.

2.4.0
- Upgraded to manifest version 2.
- Removed color picker. Now build into chrome.
- Removed ads.

2.3.2
- Add preview arrow to results on fetched pages.

2.3.1
- Shows a welcome message with choice to disable ads on install.

2.3.0
- Background color showing result of ping now configurable.
- Ping now works on all input elements and not only those of type url.
- New option to show color picker on input elements of type color.

2.2.0
- Ping check on input urls.

2.1.0
- Different colors for each digit in page number.
- Video, image and left thumbs now loads.
- Progress indicator.

js/xpath.js
/**
 * Software License Agreement (BSD License)
 * 
 * Copyright (c) 2007, Parakey Inc.
 * All rights reserved.
 * 
 * Redistribution and use of this software in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 * 
 * * Redistributions of source code must retain the above
 *   copyright notice, this list of conditions and the
 *   following disclaimer.
 * 
 * * Redistributions in binary form must reproduce the above
 *   copyright notice, this list of conditions and the
 *   following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 * 
 * * Neither the name of Parakey Inc. nor the names of its
 *   contributors may be used to endorse or promote products
 *   derived from this software without specific prior
 *   written permission of Parakey Inc.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
/**
* Gets an XPath for an element which describes its hierarchical location.
*/
this.getElementXPath = function(element)
{
    if (element && element.id)
        return '//*[@id="' + element.id + '"]';
    else
        return this.getElementTreeXPath(element);
};

this.getElementTreeXPath = function(element) {
    var paths = [];

    // Use nodeName (instead of localName) so namespace prefix is included (if any).
    for (; element && element.nodeType == 1; element = element.parentNode) {
        var index = 0;
        for (var sibling = element.previousSibling; sibling; sibling = sibling.previousSibling) {
            // Ignore document type declaration.
            if (sibling.nodeType == Node.DOCUMENT_TYPE_NODE)
                continue;

            if (sibling.nodeName == element.nodeName)
                ++index;
        }

        var tagName = element.nodeName.toLowerCase();
        var pathIndex = (index ? "[" + (index + 1) + "]" : "");
        paths.splice(0, 0, tagName + pathIndex);
    }

    return paths.length ? "/" + paths.join("/") : null;
}

this.getElementsByXPath = function(doc, xpath) {
    var nodes = [];

    try {
        var result = doc.evaluate(xpath, doc, null, XPathResult.ANY_TYPE, null);
        for (var item = result.iterateNext(); item; item = result.iterateNext())
            nodes.push(item);
    }
    catch (exc) {
        // Invalid xpath expressions make their way here sometimes.  If that happens,
        // we still want to return an empty set without an exception.
    }

    return nodes;
};


js/options.js
// Saves options to localStorage.
function save_options() {
    var enableInfiniteScroll = $("#enableInfiniteScroll").is(":checked");
    var enableTabs = $("#enableTabs").is(":checked");
    var enableFocus = $("#enableFocus").is(":checked");
    var enablePing = $("#enablePing").is(":checked");
    var enableLinks = $("#enableLinks").is(":checked");
    try {
        // Infinite scroll
        localStorage["enable_infinite_scroll"] = enableInfiniteScroll;

        // Tabs
        localStorage["enable_tabs"] = enableTabs;

        // Focus
        localStorage["enable_focus"] = enableFocus;

        // Ping
        localStorage["enable_ping"] = enablePing;

        // Links
        localStorage["enable_links"] = enableLinks;

        var message = "Options saved." + (before != after ? " A restart of Chrome is required for your new options to take effect" : "");
    } catch (e) {
        console.log(e);
    }
}

// Restores select box state to saved value from localStorage.
function restore_options() {
    var enableInfiniteScroll = localStorage["enable_infinite_scroll"];
    var enableTabs = localStorage["enable_tabs"];
    var enableFocus = localStorage["enable_focus"];
    var enablePing = localStorage["enable_ping"];
    var enableLinks = localStorage["enable_links"];

    $("#enableInfiniteScroll").attr("checked", enableInfiniteScroll == "true");
    $("#enableTabs").attr("checked", enableTabs == "true");
    $("#enableFocus").attr("checked", enableFocus == "true");
    $("#enablePing").attr("checked", enablePing == "true");
    $("#enableLinks").attr("checked", enableLinks == "true");
}

$(document).ready(function () {
    restore_options();
    $('#saveButton').click(function () {
        save_options();
        return false;
    });
    $('#cancelButton').click(function () {
        restore_options();
        return false;
    });
});


js/gInfinate.js
//TODO: support more than 10 pages
$(document).ready(function () {
    var loadcounter = 0;
    alreadyloading = false;
    $(window).scroll(function () {
        var rso = $('#rso');
        if (!rso) return;
        //alert(($(window).scrollTop() + $(window).height()) >= ($('body').height() * 0.9));
        if (($(window).scrollTop() + $(window).height()) >= ($('body').height() * 0.9)) {
            if (alreadyloading == false) {
                alreadyloading = true;
                var td = $('td[class=cur]');
                var href = $(td.siblings().get(loadcounter + 1)).find('a:first-child').attr('href');
                $.get('https://www.google.com' + href, function (data) {
                    var toAppend = $(data).find('#rso').html();
                    rso.append(toAppend);
                    alreadyloading = false;
                    loadcounter++;
                });
            }
        }
    });
});

js/background.js
///////// BOOTSTRAP ////////

function install_notice() {
    if (localStorage.getItem('install_time'))
        return;

    var now = new Date().getTime();
    localStorage.setItem('install_time', now);
    chrome.tabs.create({ url: "options.html" });
}
install_notice();

if (!localStorage["enable_focus"]) localStorage["enable_focus"] = "true"
if (!localStorage["enable_infinite_scroll"]) localStorage["enable_infinite_scroll"] = "true"
if (!localStorage["enable_links"]) localStorage["enable_links"] = "true"
if (!localStorage["enable_tabs"]) localStorage["enable_tabs"] = "false"
if (!localStorage["enable_ping"]) localStorage["enable_ping"] = "false"

function globalEval(src, callback) {
  if (window.execScript) {window.execScript(src);if (callback){callback();} return;} 
  var fn = function() {window.eval.call(window,src);
  if (callback){callback();}};fn();
}

//////// FOCUS ////////////

var xpath;

// Set focus on element in xpath.
function focusThis() {
    return function (info, tab) {
        if (!xpath) {
            return;
        }

        try {
            localStorage['focus_' + info.pageUrl] = xpath;
        } catch (e) {
            console.log(e);
            if (e == QUOTA_EXCEEDED_ERR) {
                var notification = webkitNotifications.createNotification(
                    'images/infen_logo_48x48.png',
                    'Whoops!',  // notification title
                    'Quota exceeded. Try unfocusing some sites to free up space.'
                );
                notification.show();
            }
        }
    };
}

// Remote focus for the current url.
function unfocusThis() {
    return function (info, tab) {
        localStorage.removeItem('focus_' + info.pageUrl);
    }
}

function setFocus(url) {
    var elementXpath = localStorage['focus_' + url];
    if (!elementXpath) {
		return;
    }

    chrome.tabs.getSelected(null, function(tab) {
        chrome.tabs.sendMessage(tab.id, { xpath: elementXpath }, function(response) { });
    });
}

var focusId = chrome.contextMenus.create({ "title": "Focus", "type": "normal", "contexts": ["editable"], "onclick": focusThis() });
var unfocusId = chrome.contextMenus.create({ "title": "Unfocus", "type": "normal", "contexts": ["editable"], "onclick": unfocusThis() });

////////// SHARED //////////////

chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
    if (request.method == "setFocus") {
        chrome.tabs.getSelected(null, function (tab) {
            setFocus(tab.url);
        });
    } else if (request.method == "setXPath") {
        xpath = request.xpath;
    } else if (request.method == "getLocalStorage") {
        sendResponse({ data: localStorage[request.key] });
    } else {
        sendResponse({});
    }
});


js/gInfinity-tab.js
function getIndex(input) {
    var index = -1, i = 0, found = false;
    while (i < input.form.length && index == -1)
        if (input.form[i] == input) index = i;
        else i++;
    return index;
}

var codes = new Array(8, 9, 13, 16, 17, 18, 19, 20, 27, 33, 34, 35, 36, 37, 38, 39,
                      40, 45, 46, 91, 92, 93, 112, 113, 114, 115, 116, 117, 118,
                      119, 120, 121, 122, 123, 144, 145);

chrome.extension.sendRequest({ method: "getLocalStorage", key: "enable_tabs" }, function (response) {
    if (response.data == "true") {
        document.onkeyup = function (event) {
            if ($.inArray(event.keyCode, codes) == -1) {
                if ($(event.target).is('input')) {
                    if (event.target.maxLength) {
                        if (event.target.value.length >= event.target.maxLength) {
                            var index = (getIndex(event.target) + 1);
                            var mod = index % event.target.form.length;
                            var ctrl = event.target.form[mod];
                            ctrl.focus();
                        }
                    }
                }
            }
        } 
    }
});


js/gInfinity-focus.js
chrome.extension.sendRequest({ method: "getLocalStorage", key: "enable_focus" }, function (response) {
    if (response.data == "true") {
        document.addEventListener("contextmenu", function(event) {
            if ($(event.target).is(':input')) {
                var xpath = getElementXPath(event.target);
                if (!xpath || xpath == '') {
                    return;
                }

                chrome.extension.sendRequest({ xpath: xpath, method: "context" }, function(response) { });
            }
        });
    }
});

chrome.extension.onRequest.addListener(function(request, sender, sendResponse) {

    var elements = getElementsByXPath(document, request.xpath);
    if (elements && elements.length > 0) {
        $(elements[0]).focus();
    }
    sendResponse();
});

$(document).ready(function() {
    chrome.extension.sendRequest({ method: "getLocalStorage", key: "enable_tabs" }, function (response) {
        if (response.data == "true") {
            chrome.extension.sendRequest({ method: "ready" }, function(response) {
            });
        }
    });
});


js/gInfinity.js
// Variables
var codes = new Array(8, 9, 13, 16, 17, 18, 19, 20, 27, 33, 34, 35, 36, 37, 38, 39,
                      40, 45, 46, 91, 92, 93, 112, 113, 114, 115, 116, 117, 118,
                      119, 120, 121, 122, 123, 144, 145);

reg = /(\b(https?|ftp):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gim;

function getIndex(input) {
    var index = -1, i = 0, found = false;
    while (i < input.form.length && index == -1)
        if (input.form[i] == input) index = i;
        else i++;
    return index;
}

// Listen for scroll and do infinite scroll in Google search results.
function execute() {
    alreadyloading = false;
    var nextAnchor = $('#pnnext');
    if (!nextAnchor) return;
    var href = nextAnchor.attr('href');
    var nextPage = 2;
    var previousUrl;
    $(window).scroll(function () {
        var rso = $('#rso');
        if (!rso) return;
        if (($(window).scrollTop() + $(window).height()) >= ($('body').height() * 0.9)) {
            if (alreadyloading == false) {
                alreadyloading = true;
                var parts = window.location.href.split('/');
                if (parts.length < 3) {
                    alreadyloading = false;
                    return;
                }

                if (href) {
                    rso.append($('<img/>').attr('src', chrome.runtime.getURL('images/progress.gif')).attr('id', 'ginfinity-progress').attr('style', 'margin-bottom: 20px;'));
                    $.get(href, function (data) {
                        if (document.location.href != previousUrl) {
                            nextPage = 2; // reset
                        } else {
                            nextPage++;
                        }

                        previousUrl = document.location.href;
                        var jData = $(data);
                        var toAppend = jData.find('#rso').html();
                        nextAnchor = jData.find('#pnnext');
                        if (!nextAnchor) {
                            href = null;
                        } else {
                            href = nextAnchor.attr('href');
                        }

                        // Insert preview in all results
                        while (toAppend.indexOf('</h3><div class="s">') != -1) {
                            toAppend = toAppend.replace('</h3><div class="s">', '</h3><div class="vspib" aria-label="Result details" role="button" tabindex="0"><div class="vspii"><div class="vspiic"></div></div></div><div class="s">');
                        }

                        /// Show potential video thumbs
                        var scripts = new Array();
                        $.each($(toAppend).find('img[id^="vidthumb"],img[id^="imgthumb"],img[id^="leftthumb"]'), function (i, v) {
                            var random = Math.floor(Math.random() * 1000000000);
                            var id = $(this).attr('id');
                            toAppend = toAppend.replace('id="' + id, 'id="' + random);
                            jData.filter('script').each(function (e, v) {
                                var html = $(v).html();
                                if (html != null && html.indexOf("(function(x){x&&(x.src='data:image/jpeg;base64") != -1 && html.indexOf(id) != -1) {
                                    scripts.push(html.replace(id, random));
                                }
                            });
                        });

                        var colors = ["007D08", "D62408", "1645AE"];
                        var pageHeader = '<span style="text-shadow:1px 1px 2px #000;font-weight:bold;font-size:x-large;"><span style="color:#1645AE;">P</span><span style="color:#D62408;">a</span><span style="color:#EFBA00;">g</span><span style="color:#1645AE;">e</span> ';
                        $.each(nextPage.toString().split(""), function (i, v) {
                            pageHeader = pageHeader + '<span style="color:#' + colors[i] + ';">' + v + '</span>';
                        });
                        pageHeader = pageHeader + '</span>';

                        $('#ginfinity-progress').remove();
                        rso.append($('<li></li>').attr('class', 'g').html(pageHeader));
                        rso.append(toAppend);

                        $.each(scripts, function (i, v) {
                            eval(v);
                        });

                        alreadyloading = false;
                    });
                } else {
                    alreadyloading = false;
                }
            }
        }
    });
}

chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
    var elements = getElementsByXPath(document, request.xpath);
    if (elements && elements.length > 0) {
        $(elements[0]).focus();
    }
    sendResponse();
});

// Send xpath info about selected input field when clicking the context menu.

chrome.runtime.sendMessage({ method: "getLocalStorage", key: "enable_focus" }, function (response) {
    if (response.data == "true") {
        document.addEventListener("contextmenu", function (event) {
            if ($(event.target).is(':input')) {
                var xpath = getElementXPath(event.target);
                if (!xpath || xpath == '') {
                    return;
                }

                chrome.runtime.sendMessage({ xpath: xpath, method: "setXPath" }, function (response) { });
            }
        });
    }
});

$(document).ready(function () {
    if (document.location.href.indexOf('://www.google.') != -1) {
        chrome.runtime.sendMessage({ method: "getLocalStorage", key: "enable_infinite_scroll" }, function (response) {
            if (response.data == "true") {
                // Wait for Google to load the page
                setTimeout('execute()', 1000);
            }
        });
    }

    chrome.runtime.sendMessage({ method: "getLocalStorage", key: "enable_focus" }, function (response) {
        if (response.data == "true") {
            chrome.runtime.sendMessage({ method: "setFocus" }, function (response) {
            });
        }
    });

    chrome.runtime.sendMessage({ method: "getLocalStorage", key: "enable_links" }, function (response) {
        if (response.data == "true" && document.getElementById && document.createTreeWalker && typeof NodeFilter != "undefined") {
            var tw = document.createTreeWalker(document.body, NodeFilter.SHOW_TEXT, null, false);
            var node;
            var rem = [];

            while (node = tw.nextNode()) {
                if (node.nodeValue.match(reg) && node.parentNode.tagName != 'A' && node.parentNode.tagName != 'TEXTAREA' && node.parentNode.tagName != 'STYLE' && node.parentNode.tagName != 'SCRIPT' && node.parentNode.tagName != 'META' && node.parentNode.tagName != 'NOSCRIPT' && node.parentNode.tagName != 'PRE') {
                    var parentElem = node.parentNode;
                    var val = node.nodeValue.replace(reg, '<a href="$1">$1</a>');
                    var aNode = document.createElement("span");
                    aNode.innerHTML = val;
                    parentElem.insertBefore(aNode, node);
                    rem.push(node);
                    continue;
                }
            }
            for (i in rem) {
                rem[i].parentNode.removeChild(rem[i]);
            }
        }
    });

    // Assign keyup listener if tabs feature enabled.

    chrome.runtime.sendMessage({ method: "getLocalStorage", key: "enable_tabs" }, function (response) {
        if (response.data == "true") {
            document.onkeyup = function (event) {
                if ($.inArray(event.keyCode, codes) == -1) {
                    if ($(event.target).is('input')) {
                        if (event.target.maxLength > 0) {
                            if (event.target.value.length >= event.target.maxLength) {
                                var index = (getIndex(event.target) + 1);
                                var mod = index % event.target.form.length;
                                var ctrl = event.target.form[mod];
                                ctrl.focus();
                            }
                        }
                    }
                }
            }
        }
    });

    chrome.runtime.sendMessage({ method: "getLocalStorage", key: "enable_ping" }, function (response) {
        if (response.data == "true") {
            $('input').blur(function () {
                var input = this;
                var val = $(input).val();
                if (val && val != null && val.match(reg)) {
                    $.ajax(val).fail(function () {
                        $(input).css('background', '#FFDDDD');
                    }).done(function () {
                        $(input).css('background', '#DCFBD1');
                    });
                }
            });
        }
    });
});
